File: src/engine/badges.ts
// Puzzle badges: self-descriptive traits that hint at how a puzzle will feel to
// solve. Badges that make a puzzle easier reduce how much it contributes to the
// overall Clueweave Score (per-puzzle 0–100 scores stay normalized); a badge
// with a multiplier above 1 would make a puzzle count for more.

import type { Puzzle } from "./types";
import { detectSymmetry } from "./symmetry";

export type BadgeKey = "symmetric" | "named" | "patterned";

export interface Badge {
  key: BadgeKey;
  /** Chip text, e.g. "◈ Symmetric · H+V". */
  label: string;
  /** Weight applied to this puzzle's share of the Clueweave Score. <1 = easier. */
  multiplier: number;
  /** One-line explanation shown on the badge's filter page. */
  blurb: string;
}

export const BADGE_INFO: Record<BadgeKey, { name: string; icon: string; multiplier: number; blurb: string }> = {
  symmetric: {
    name: "Symmetric",
    icon: "◈",
    multiplier: 0.85,
    blurb: "The picture mirrors itself, so every deduction on one side gives you the other side for free.",
  },
  named: {
    name: "Name-hint",
    icon: "🏷",
    multiplier: 0.9,
    blurb: "The title tells you what you're drawing, so you can often guess where the picture is headed.",
  },
  patterned: {
    name: "Patterned",
    icon: "▤",
    multiplier: 0.8,
    blurb: "Every row and column is one solid run — once the shape starts, you mostly continue the pattern.",
  },
};

/** True when every row AND every column contains at most one run of filled
 *  cells — the shape is a single "convex-ish" block you can continue. */
export function detectPatterned(grid: boolean[][]): boolean {
  const runs = (line: boolean[]): number => {
    let n = 0;
    let inRun = false;
    for (const v of line) {
      if (v && !inRun) {
        n++;
        inRun = true;
      } else if (!v) {
        inRun = false;
      }
    }
    return n;
  };
  if (grid.length === 0) return false;
  if (grid.some((row) => runs(row) > 1)) return false;
  for (let c = 0; c < grid[0].length; c++) {
    if (runs(grid.map((r) => r[c])) > 1) return false;
  }
  return true;
}

/**
 * Plain-English meaning of every code the Symmetric chip can show. The chip is
 * terse by necessity ("◈ Symmetric · H"), so this is the single source of truth
 * the help surfaces read from — keep it in sync with `symmetryDetail` below.
 */
export const SYMMETRY_LEGEND: Array<{ code: string; meaning: string }> = [
  { code: "H", meaning: "Mirrors left ↔ right. Fold it down the middle and the two halves match." },
  { code: "V", meaning: "Mirrors top ↔ bottom. Fold it across the middle and the two halves match." },
  { code: "H+V", meaning: "Mirrors both ways at once — left ↔ right and top ↔ bottom." },
  { code: "180°", meaning: "No mirror, but turn the picture upside-down and you get the same picture." },
];

/** Human detail for the symmetric badge: which way the picture mirrors. */
export function symmetryDetail(grid: boolean[][]): string | null {
  const s = detectSymmetry(grid);
  if (s.horizontal && s.vertical) return "H+V";
  if (s.horizontal) return "H";
  if (s.vertical) return "V";
  if (s.rotational) return "180°";
  return null;
}

/** All badges that apply to a puzzle (auto-detected + the curated name flag). */
export function puzzleBadges(puzzle: Pick<Puzzle, "solution" | "named">): Badge[] {
  const badges: Badge[] = [];
  const sym = symmetryDetail(puzzle.solution);
  if (sym) {
    const info = BADGE_INFO.symmetric;
    badges.push({
      key: "symmetric",
      label: `${info.icon} ${info.name} · ${sym}`,
      multiplier: info.multiplier,
      blurb: info.blurb,
    });
  }
  if (puzzle.named) {
    const info = BADGE_INFO.named;
    badges.push({ key: "named", label: `${info.icon} ${info.name}`, multiplier: info.multiplier, blurb: info.blurb });
  }
  if (detectPatterned(puzzle.solution)) {
    const info = BADGE_INFO.patterned;
    badges.push({ key: "patterned", label: `${info.icon} ${info.name}`, multiplier: info.multiplier, blurb: info.blurb });
  }
  return badges;
}

/** Combined Clueweave-Score weight multiplier for a puzzle's badges. */
export function badgeWeightMultiplier(badges: Badge[]): number {
  return badges.reduce((m, b) => m * b.multiplier, 1);
}

File: ClueweaveKit/Sources/ClueweaveKit/Badges.swift
// Puzzle badges (port of src/engine/badges.ts): self-descriptive traits that
// shift a puzzle's weight in the overall Clueweave Score. <1 = easier.

public enum BadgeKey: String, CaseIterable, Sendable {
    case symmetric, named, patterned

    public var name: String {
        switch self {
        case .symmetric: "Symmetric"
        case .named: "Name-hint"
        case .patterned: "Patterned"
        }
    }

    public var icon: String {
        switch self {
        case .symmetric: "◈"
        case .named: "🏷"
        case .patterned: "▤"
        }
    }

    /// A small geometric SF Symbol for the badge — used where a compact,
    /// non-clickable indicator is wanted (puzzle tiles, the watch, legends).
    public var glyph: String {
        switch self {
        case .symmetric: "rhombus.fill"
        case .named: "tag.fill"
        case .patterned: "square.grid.3x3.fill"
        }
    }

    public var multiplier: Double {
        switch self {
        case .symmetric: 0.85
        case .named: 0.9
        case .patterned: 0.8
        }
    }

    public var blurb: String {
        switch self {
        case .symmetric:
            "The picture mirrors itself, so every deduction on one side gives you the other side for free."
        case .named:
            "The title tells you what you're drawing, so you can often guess where the picture is headed."
        case .patterned:
            "Every row and column is one solid run — once the shape starts, you mostly continue the pattern."
        }
    }
}

public struct Badge: Sendable, Equatable {
    public let key: BadgeKey
    /// Chip text, e.g. "◈ Symmetric · H+V".
    public let label: String
    public var multiplier: Double { key.multiplier }
}

/// True when every row AND every column contains at most one run of filled cells.
public func detectPatterned(_ grid: [[Bool]]) -> Bool {
    func runs(_ line: [Bool]) -> Int {
        var n = 0
        var inRun = false
        for v in line {
            if v && !inRun {
                n += 1
                inRun = true
            } else if !v {
                inRun = false
            }
        }
        return n
    }
    guard !grid.isEmpty else { return false }
    if grid.contains(where: { runs($0) > 1 }) { return false }
    for c in 0..<grid[0].count {
        if runs(grid.map { $0[c] }) > 1 { return false }
    }
    return true
}

/// Plain-English meaning of every code the Symmetric chip can show. The chip is
/// terse by necessity ("◈ Symmetric · H"), so this is the single source of truth
/// the help surfaces read from — keep it in sync with `symmetryDetail` below.
public struct SymmetryCode: Sendable, Equatable {
    public let code: String
    public let meaning: String
}

public let symmetryLegend: [SymmetryCode] = [
    SymmetryCode(
        code: "H",
        meaning: "Mirrors left ↔ right. Fold it down the middle and the two halves match."),
    SymmetryCode(
        code: "V",
        meaning: "Mirrors top ↔ bottom. Fold it across the middle and the two halves match."),
    SymmetryCode(
        code: "H+V",
        meaning: "Mirrors both ways at once — left ↔ right and top ↔ bottom."),
    SymmetryCode(
        code: "180°",
        meaning: "No mirror, but turn the picture upside-down and you get the same picture."),
]

/// Human detail for the symmetric badge: which way the picture mirrors.
public func symmetryDetail(_ grid: [[Bool]]) -> String? {
    let s = detectSymmetry(grid)
    if s.horizontal && s.vertical { return "H+V" }
    if s.horizontal { return "H" }
    if s.vertical { return "V" }
    if s.rotational { return "180°" }
    return nil
}

/// All badges that apply to a puzzle (auto-detected + the curated name flag).
public func puzzleBadges(solution: [[Bool]], named: Bool) -> [Badge] {
    var badges: [Badge] = []
    if let detail = symmetryDetail(solution) {
        badges.append(Badge(key: .symmetric, label: "◈ Symmetric · \(detail)"))
    }
    if named {
        badges.append(Badge(key: .named, label: "🏷 Name-hint"))
    }
    if detectPatterned(solution) {
        badges.append(Badge(key: .patterned, label: "▤ Patterned"))
    }
    return badges
}

public func puzzleBadges(_ puzzle: Puzzle) -> [Badge] {
    puzzleBadges(solution: puzzle.solution, named: puzzle.named)
}

/// Combined Clueweave-Score weight multiplier for a puzzle's badges.
public func badgeWeightMultiplier(_ badges: [Badge]) -> Double {
    badges.reduce(1) { $0 * $1.multiplier }
}
